sqlglot/dialects/bigquery.py

```diff
@@ -591,15 +591,15 @@ def _parse_unnest(self, with_alias: bool = True) -> t.Optional[exp.Unnest]:
 
             if unnest:
                 expr = seq_get(unnest.expressions, 0)
+                if expr:
+                    from sqlglot.optimizer.annotate_types import annotate_types
+
+                    expr = annotate_types(expr)
 
-                # Unnesting a nested array (i.e array of structs) in BQ explodes the struct fields,
-                # in contrast to other dialects such as DuckDB which flattens only the array by default
-                if (
-                    expr
-                    and (isinstance(expr, exp.Array) or expr.is_type(exp.DataType.Type.ARRAY))
-                    and expr.find(exp.Struct)
-                ):
-                    unnest.set("nested_array", True)
+                    # Unnesting a nested array (i.e array of structs) in BQ explodes the top-level struct fields,
+                    # in contrast to other dialects such as DuckDB which flattens only the array by default
+                    if expr.is_type(exp.DataType.Type.ARRAY) and expr.find(exp.Struct):
+                        unnest.set("explode_array", True)
```

Comment:
Calling `expr.find` is not very robust because it searches the whole tree and may find a struct in a lower depth. Also, with this approach we don't leverage the inferred types at all, so we may miss functions / complex expressions that produce structs but aren't `exp.Struct`.

Consider instead breaking up the condition and being more explicit:

1. check if `is_type(array)`
2. check if `expressions[0].is_type` struct, unless there are no expressions set in the array